urbit/app/eth-event-viewer.hoon

```diff
@@ -0,0 +1,409 @@
+++  ux-to-cord
+  |=  address=@ux
+  ^-  @t
+::  crip transforms tape to cord
+  (cat 3 '0x' (crip ((x-co:co 4) address)))
+::
+++  to-eth-watcher
+  |=  [=wire =task:agent:gall]
+  ^-  card
+  [%pass wire %agent [our.bol %eth-watcher] task]
+::
+++  setup-eth-watcher
+  |=  contract=contract-type
+  =/  url  'http://eth-mainnet.urbit.org:8545'
+  %+  to-eth-watcher  (get-path address.contract)
+  :+  %poke   %eth-watcher-poke
+  !>  ^-  poke:ew-sur
+  :+  %watch  (get-path address.contract)
+  :*  url
+      |
+      ~m2
+      ~m2
+      9.960.593
+      ~[address.contract]
+      (get-topics specific-events.contract)
+  ==
+::
+++  watch-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  (get-path address)
+  [%watch (get-logs-path address)]
+::
+++  leave-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  (get-path address)
+  [%leave ~]
+::
+++  clear-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  /clear
+  :+  %poke  %eth-watcher-poke
+  !>  ^-  poke:ew-sur
+  [%clear (get-path address)]
+::
+++  get-topics
+  |=  specific-events=(list @t)
+  ^-  (list ?(@ux (list @ux)))
+  ?~  specific-events
+    ~
+  :~  `(list @ux)`(turn specific-events hex-list)
+  ==
+::
+++  hex-list  |=(e=@t `@ux`(event-string-to-hex (trip e)))
+::
+++  event-string-to-hex
+  |=  event-string=tape
+  ^-  @ux
+  `@ux`(keccak-256:keccak:crypto (as-octt:mimes:html event-string))
+::
+++  get-path
+  |=  address=@ux
+  ^-  path
+  `path`/[dap.bol]/(scot %tas (ux-to-cord address))
+::
+++  get-logs-path
+  |=  address=@ux
+  ^-  path
+  `path`/logs/[dap.bol]/(scot %tas (ux-to-cord address))
+::
+++  contract-cord-to-hex
+  |=  address=@t
+  ^-  @ux
+  =/  address-tape
+    (cass q:(trim 2 (trip address)))
+  `@ux`(scan address-tape hex)
+::
+++  poke-action
+  |=  action=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?-  -.action
+      %add-contract  (handle-add-contract action)
+      %get-abi  (handle-get-abi action)
+      %remove-contract  (handle-remove-contract action)
+      %reload-events  (handle-reload-events action)
+  ==
+::
+++  handle-reload-events
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%reload-events -.act)
+  :_  state
+  :~  (leave-eth-watcher address.act)
+      (watch-eth-watcher address.act)
+  ==
+::
+++  handle-get-abi
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%get-abi -.act)
+  :_  state
+  [(request-ethereum-abi address.act) ~]
+::
+++  handle-add-contract
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%add-contract -.act)
+  ?:  =(%.y (~(has by contracts.state) address.contract.act))
+      [~ state]
+  =/  address  address.contract.act
+  :_  state(contracts (~(put by contracts.state) address contract.act))
+  :~  (request-ethereum-abi address.contract.act)
+      [%give %fact [/state/update ~] %json !>((new-json contract.act))]
+      (setup-eth-watcher contract.act)
+      (watch-eth-watcher address.contract.act)
+  ==
+::
+++  new-json
```

Comment:
These json conversion functions might find a better home in the `/lib` file for the app.